### ISSUE TITLE
mWeb (Safari) - Dragging the screen up creates empty space between keyboard and composer - Reported by: @parasharrajat 

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#5894]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Navigate to a conversation 
2. Tap on the composer box 
3. Drag the screen upwards

## Expected Result:
Dragging the screen up should not show any blank space

## Actual Result:
Dragging the screen up shows a blank space

## Workaround:
None needed, visual issue. 

## Platform:

Where is this issue occurring?

- Mobile Web (Safari) 

**Version Number:** 1.1.8-0

**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation
![image](https://user-images.githubusercontent.com/44479856/137534709-d588d153-18a7-40bf-8c05-eed411ba9ab6.png)

**Expensify/Expensify Issue URL:**

**Issue reported by:** @parasharrajat 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1633858525070700

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Dal-Papa (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Christinadobrzyn (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @roryabraham (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by Christinadobrzyn** [comment]
@parasharrajat are you still interested in this job?

**Comment by parasharrajat** [comment]
Haven't looked at it yet. If I have a solution, I will propose here. Thanks for asking. 🙂

**Comment by sidferreira** [comment]
@Christinadobrzyn Not sure if fixable: https://stackoverflow.com/questions/61014250/empty-space-on-the-bottom-of-the-page-when-swipe-up-on-safari-ios . I tried the same as https://github.com/Expensify/App/issues/5892#issuecomment-960382897 but didn't work.

**Comment by Christinadobrzyn** [comment]
Raised price $500 for the job.

**Comment by Christinadobrzyn** [comment]
Thanks for the insight @sidferreira - @roryabraham do you have any suggestions for that?

**Comment by roryabraham** [comment]
I don't think that the answer provided in https://stackoverflow.com/questions/61014250/empty-space-on-the-bottom-of-the-page-when-swipe-up-on-safari-ios is relevant in this case because the white space persists after you lift your finger (i.e: it doesn't "bounce back")

**Comment by sidferreira** [comment]
@roryabraham I may dive on it over the weekend... had a couple of ideas

**Comment by mallenexpensify** [comment]
We are placing all non-critical issues on hold while we're on a company offsite.  The hold is expected to be lifted on 11/19 (but could go longer).  For any PRs that are submitted during the hold, we will add a $250 bonus payment.

**Comment by roryabraham** [comment]
@mallenexpensify @Christinadobrzyn I think we are good to take this off hold?

**Comment by Christinadobrzyn** [comment]
Sounds good - still looking for proposals

**Comment by Christinadobrzyn** [comment]
Raised price to $1000 - still waiting on proposals

**Comment by justsj11** [comment]
How can I get access to your stack overflow? I have been looking to get into your codebase/community, but is unable to do so, as they require a expensify email everywhere :(

**Comment by mallenexpensify** [comment]
Hi @justsj11 , Stack Overflow is currently only accessible to internal employees.  Have you reviewed https://github.com/Expensify/App/blob/main/CONTRIBUTING.md and https://github.com/Expensify/App/blob/main/README.md ?  Those should help answer many questions. Also, if you haven't done so yet, you can request access to our #expensify-open-source Slack channel by emailing contributors@expensify.com

**Comment by sidferreira** [comment]
@roryabraham @Christinadobrzyn so, after further investigation, reading https://blog.opendigerati.com/the-eccentric-ways-of-ios-safari-with-the-keyboard-b5aa3f34228d, lots of stack overvflow, finding out that it is NOT the app, I actually found a workaround.

The solution includes using pure javascript to modify the browser behavior.
Basically:
```
function preventDefault(e){
    e.preventDefault();
}

function disableScroll(){
    document.body.classList.add('padding-bottom');
    document.body.addEventListener('touchmove', preventDefault, { passive: false });
}
function enableScroll(){
    document.body.classList.remove('padding-bottom');
    document.body.removeEventListener('touchmove', preventDefault);
}
```

this code would:
a) just prevent scrolling when the keyboard is focused
b) add a ~30px padding at the bottom of the screen to make sure the [new address bar won't cover the input](https://user-images.githubusercontent.com/143615/142517879-029f6596-94fd-4810-8db7-e72dea9ee9bd.png)

This should be applied only on iOS devices.

**Comment by roryabraham** [comment]
@sidferreira So to clarify your proposal, when the main chat composer is focused on iOS Safari, then we'll disable scrolling on the entire document. Then we'll re-enable it when the composer is blurred. Is that correct?

If so, I'd say that's unfortunately not an acceptable solution, but could we instead disable scrolling on the composer itself?

**Comment by sidferreira** [comment]
@roryabraham it disables the scroll on the window it self. Inner components, with scroll, should still work. Safari adds a ~300px area at the bottom of the document, which we can barely identify. 

I could try to control the amount of scroll that the user does, and if it generates the blank space, I can smoothly scroll back... Basically code the behaviour we expected to have. This solution requires a bit more JS.

**Comment by roryabraham** [comment]
> @roryabraham it disables the scroll on the window it self. Inner components, with scroll, should still work

I see, thanks for clarifying.

> I could try to control the amount of scroll that the user does, and if it generates the blank space, I can smoothly scroll back... Basically code the behaviour we expected to have. This solution requires a bit more JS.

I think that would be ideal to maintain a consistent experience, but it's not a hard requirement.

One other question I have is: is this problem universal with react-native-web? If we can reproduce the issue consistently with a minimal expo snack, I wonder if the maintainer of react-native-web would be open to implementing the solution directly in react-native-web. Let's focus on getting our app working first, but I could see this being a valuable gift to the open-source community if we get an elegant solution.

Thanks for all your investigation @sidferreira 👍 I am going to ask for a buddy-check here before green-lighting your solution, but it sounds good to me.

**Comment by sidferreira** [comment]
@roryabraham I manage to cause the issue it with `react-boilerplate`

**Comment by aswin-s** [comment]
The is indeed an iOS quirk like @sidferreira mentioned as the scroll is outside the html element when virtual keyboard is opened. So there is no way to prevent the scroll from code. Browser simply changes the inset of the content to position the focused input with in the viewable area. Surprised to see this bug even on sites like https://bing.com.

However we could mitigate this issue by dismissing the virtual keyboard when someone taps outside the TextInput. That's how some of the popular mobile apps have solved this issue. Checkout https://duckduckgo.com for example. My proposal is to make use of the `<SwippableView>` HOC around `ReportActionCompose`. Right now `SwippableView` lacks an implementation for web platform. Instead use it to capture touchstart & drag events outside out textinput and blur the input if the event target was not an `input` element. This would dismiss the virtual keyboard if user attempts to swipe or interact with elements outside of input element.

**Comment by roryabraham** [comment]
Thanks for the discussion @sidferreira @aswin-s!

> However we could mitigate this issue by dismissing the virtual keyboard when someone taps outside the TextInput. That's how some of the popular mobile apps have solved this issue. Checkout duckduckgo.com for example

I was skeptical of this approach, but it actually works pretty well on duckduckgo.com. One minor correction though – we'd only want to dismiss the virtual keyboard if someone taps on the view wrapping the TextInput I believe. We still want the main `ReportActionsView` to be scrollable while the virtual keyboard is open. Also this approach fails if the user taps _on_ the text input and scrolls up – then we end up with the same issue.

So to summarize our options so far, we have:

1. Use javascript to disable scrolling on the window when the `ReportActionCompose` is focused.

     - Pros: eliminates the issue with a simple solution
     - Cons: removes the "pull to refresh" style of benign interactivity that iOS users are used to

1. Allow scrolling on the window, but measure the amount of scrolling the user does and smoothly scroll back.

    - Pros: eliminates the issue, maintains the behavior we would've expected to be the default and which is intuitive to users.
    - Cons: more complex solution.
    - Additional note: might be worth seeing if we can implement this generic solution directly in react-native-web to help out others with this (apparently common) problem and to prevent the slightly more complex code from having to live in this repo.

1. Dismiss the keyboard if the user attempts to tap outside the TextInput.

    - Pros: Simplest solution, no hacky javascript. Could implement uniformly across platforms (i.e: doesn't necessarily introduce Safari-specific code)
    - Cons: Doesn't work if the user taps on the TextInput and scrolls. Needs clarification on how it would allow scrolling of main `ReportActionsView` while keyboard is open. Also is maybe the least premium experience for the user imo.
    
Thoughts?

**Comment by tgolen** [comment]
Out of those options, I like them in this order: 2, 1, 3. I think it would be good for us to explore a proposal for 2 so we can better understand what the complexity would look like, and ideally, working it into react-native-web would be great!

**Comment by sidferreira** [comment]
@roryabraham @tgolen I believe it is possible to make it as a separate js, in a way it can be exported and reused in other projects. And, as always, keeping it as vanilla as possible.

**Comment by roryabraham** [comment]
Okay, sounds good. @sidferreira let's get you hired on Upwork so you can get started working on solution `2` above, and get back to us! @Christinadobrzyn please hire @sidferreira on Upwork.

**Comment by sidferreira** [comment]
@roryabraham @tgolen Follows a small sample of what I got so far: https://www.dropbox.com/s/zqyc2eokcsewyxe/Screen%20Recording%202021-11-21%20at%208.02.08%20PM.mov?dl=0

I still have to test on some different devices.
I also had to use `https://github.com/iamdustan/smoothscroll` but it would be required only for web and loaded only for Safari, so I believe it isn't an issue (only 4kb)

**Comment by Christinadobrzyn** [comment]
Hired @sidferreira in Upwork 🎉

**Comment by roryabraham** [comment]
Thanks for the update @sidferreira

**Comment by sidferreira** [comment]
@roryabraham @tgolen https://github.com/Expensify/App/pull/6413

**Comment by parasharrajat** [comment]
I am reviewing the PR and leaving a comment for visibility.  [See More](https://expensify.slack.com/archives/C01GTK53T8Q/p1637728779426100)

**Comment by roryabraham** [comment]
PR merged, @Christinadobrzyn see [this comment](https://github.com/Expensify/App/pull/6413#issuecomment-982220624) for potential to add another milestone to this issue.

**Comment by MelvinBot** [comment]
📣 @sidferreira You have been assigned to this job by @Christinadobrzyn!
Please apply to this job in Upwork and let us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by pecanoro** [comment]
As mentioned here, there is a regression in this issue that needs a follow-up PR https://github.com/Expensify/App/issues/6621

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.18-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/6413


If no regressions arise, payment will be issued on 2021-12-15. :confetti_ball:

**Comment by Christinadobrzyn** [comment]
7 day production timeframe is complete. I'm working on payment for this - just checking with the team about the Upwork payment structure. I'll follow up as soon as possible!

**Comment by parasharrajat** [comment]
There is an open regression @Christinadobrzyn. so we should wait for that.

**Comment by Christinadobrzyn** [comment]
Hi @parasharrajat can you confirm this is the [PR](https://github.com/Expensify/App/pull/6572) to fix the regression so I can track it for payment?

**Comment by parasharrajat** [comment]
Yes @Christinadobrzyn

**Comment by laurenreidexpensify** [comment]
@parasharrajat can you confirm what latest is here? Is this ready for payment now?

**Comment by parasharrajat** [comment]
Not yet ready for payment.

**Comment by roryabraham** [comment]
@sidferreira has been working diligently to try and get this fixed properly. We might end up having to revert the initial solution and instead do a simpler one where we just disable scrolling on the composer.

**Comment by sidferreira** [comment]
@roryabraham I think I should:

a) create a PR reverting the changes
b) create another PR trying https://github.com/Expensify/App/issues/5894#issuecomment-973563994 ?

**Comment by parasharrajat** [comment]
@roryabraham What do you think should we do this? Preventing scroll is fine but I am not sure about adding bottom padding. I can't imagine the UI without trying out this solution. Do you think we should give it a try?

**Comment by parasharrajat** [comment]
@roryabraham Any thoughts?

**Comment by mallenexpensify** [comment]
@roryabraham , when you have a min can you check Rajat's and Sid's comments above?

**Comment by roryabraham** [comment]
Okay, sorry for the delay here everyone. I think that @sidferreira's solution proposed [here](https://github.com/Expensify/App/issues/5894#issuecomment-973563994) makes sense – though how will we know the correct height for the address bar? It's not always the same IIRC.

**Comment by sidferreira** [comment]
@roryabraham TBH I'm starting to think if this even have a good solution. I had some recent ideas, to test, but TBH I spent a lot of time already. I would like to debug and understand why some bugs happen. I wonder if it is possible to raise the reward having in mind the challenge that it offers...

**Comment by roryabraham** [comment]
Yes, this one has been a big PITA and I've been impressed with the research and work you've done so far. let's bump the budget to $1500?

**Comment by mallenexpensify** [comment]
Noted the $1500, inc in the title. I'm unable to raise the job price in Upwork cuz @sidferreira is already hired, @Christinadobrzyn will add the balance as a bonus when paying

**Comment by roryabraham** [comment]
@sidferreira It's been 10 days since you provided your last update. Can you please provide us with a status update on this issue?

**Comment by sidferreira** [comment]
@roryabraham I had some hard time past week, I hope to be able to run the tests this week. If you think is best, feel free to unassign me...

**Comment by mallenexpensify** [comment]
@sidferreira do you foresee being able to get to it this week? If so, let's leave ya assigned.  In the future, try to provide daily, or at least frequent, updates on issues.  We recently updated the CONTRIBUTING.md to include the below ([link](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#:~:text=Daily%20updates%20on%20weekdays%20are%20highly%20recommended.))


> Daily updates on weekdays are highly recommended. If you know you won’t be able to provide updates for > 1 week, please comment on the PR or issue how long you plan to be out so that we may plan accordingly. We understand everyone needs a little vacation here and there. Any issue that doesn't receive an update for 1 full week may be considered abandoned and the original contract terminated.

**Comment by sidferreira** [comment]
@mallenexpensify @roryabraham things are piling up, I really think it's the best to unassign me...

**Comment by parasharrajat** [comment]
Ok. Thanks for update @sidferreira.

If we are looking to unassign sidferreira. Then I suggest that we revert the orginal PR as it is causing issue. I would also suggest to close this issue as this is not urgent and an extreme edge case. Most of the app faced similar issues one way or another.

If we are looking to get this fixed, create a fresh issue at $250.

**Comment by roryabraham** [comment]
> @mallenexpensify @roryabraham things are piling up, I really think it's the best to unassign me...

Thanks for the transparency @sidferreira !

I agree with @parasharrajat's evaluation here with closing this issue for now. I also think @sidferreira deserves a payout for the work done in https://github.com/Expensify/App/pull/6413 for the amount this job was listed at at the time he was hired (assuming he's not been paid out already).

**Comment by Christinadobrzyn** [comment]
Sounds good! To be 100% sure @roryabraham, I can pay @sidferreira in Upwork, close the Upwork job, and this GH issue?

**Comment by parasharrajat** [comment]
@Christinadobrzyn I am not able to apply for C+ on the sent invitation. There must be some issue from Upwork. I think you will have to create another job for that. 

You may need
1. Original price of the job was $1000, not $1500. rest of the work is being canceled.
2. I reviewed the PR as a C+ https://github.com/Expensify/App/issues/5894#issuecomment-981950278. Due to the reason that C+ launched officially after this issue, the proper C+ process was not followed for this.

**Comment by roryabraham** [comment]
What @parasharrajat said is correct. @sidferreira should get $1000 and @parasharrajat should get paid for the C+ review.

**Comment by sidferreira** [comment]
@roryabraham I'm most grateful for that!

**Comment by roryabraham** [comment]
Going to unassign myself and @sidferreira here. @Christinadobrzyn feel free to close this out as soon as payments have been issued.

**Comment by Christinadobrzyn** [comment]
Thanks for clarifying @parasharrajat and @roryabraham! 

I just paid $1000 @sidferreira through the [old job post](https://www.upwork.com/ab/applicants/1450311669048020992/job-details). 

I created a [new job post](https://www.upwork.com/ab/applicants/1486940692653445120/job-details) and hired/paid @parasharrajat for the C+ and bonus for reporting this issue. I think we're good! 

Closing this issue

**Comment by roryabraham** [comment]
Reopening this.

**Comment by Christinadobrzyn** [comment]
Hey @roryabraham just catching up here, it looks like the fix for this was reverted due to subsequent bugs? Should I create a new Upwork job or are we going to fix this internally?

**Comment by roryabraham** [comment]
@Christinadobrzyn that's correct – this was reverted due to subsequent bugs. And yes, let's create a new Upwork job please 🙂

**Comment by Christinadobrzyn** [comment]
Created a new Upwork job to fix these bugs

Internal posting - https://www.upwork.com/ab/applicants/1527476784280002560/job-details
External posting - https://www.upwork.com/nx/job-post/custom/preview?uid=1527476784280002560

**Comment by Beamanator** [comment]
@roryabraham looks like I got assigned b/c you're OOO today - I'm going to unassign since i believe you'll be back Monday, feel free to add me back if you need my help 👍

**Comment by mananjadhav** [comment]
Thanks for the comment @jrstanley. @Beamanator @Christinadobrzyn We can close this out in favor of the traction in #8731. I would also like to highlight that this is reported by @parasharrajat long before the other issue and might be eligible for reporting bonus if we fix either of these issues.

**Comment by parasharrajat** [comment]
No, I was already paid reporting bonus so I am good.

**Comment by mananjadhav** [comment]
> No, I was already paid reporting bonus so I am good.

Okay thanks for the update.

**Comment by roryabraham** [comment]
> we can close this out in favor of the traction in https://github.com/Expensify/App/issues/8731.

This is fine with me as long as @luacmartins familiarizes himself with all the context in this issue and understands what's already been tried. @luacmartins if you feel confident in that, would you please close this out?

**Comment by luacmartins** [comment]
Oof lots of comments here. I'll take a look tomorrow, but it seems like this issue has more context than https://github.com/Expensify/App/issues/8731.

**Comment by roryabraham** [comment]
The main thing that's already been tried is what was reverted [here](https://github.com/Expensify/App/pull/8571), and I think the other main suggestions are summarized at a high level in [this comment](https://github.com/Expensify/App/issues/5894#issuecomment-973828800). Hopefully that helps distill the issue for you 🙂

**Comment by luacmartins** [comment]
Thanks for the summary @roryabraham, that was really helpful! Closing this in favor of https://github.com/Expensify/App/issues/8731

**Comment by Christinadobrzyn** [comment]
Cancelled the job in Upwork
